src/plugins/formulas/engine/register.js

```diff
@@ -118,14 +148,24 @@ export function getRegisteredHotInstances(engine) {
  */
 export function unregisterEngine(engine, hotInstance) {
   if (engine) {
-    const engineRegistry = staticRegister(PLUGIN_KEY).getItem(ENGINE_KEY);
-    const sharedEngineUsage = engineRegistry?.get(engine);
+    const engineRegistry = getEngineRelationshipRegistry();
+    const engineHotRelationship = engineRegistry.get(engine);
+    const sharedEngineRegistry = getSharedEngineUsageRegistry();
+    const sharedEngineUsage = sharedEngineRegistry.get(engine);
 
-    if (sharedEngineUsage && sharedEngineUsage.includes(hotInstance)) {
-      sharedEngineUsage.splice(sharedEngineUsage.indexOf(hotInstance), 1);
+    if (engineHotRelationship && engineHotRelationship.includes(hotInstance)) {
+      engineHotRelationship.splice(engineHotRelationship.indexOf(hotInstance), 1);
 
-      if (sharedEngineUsage.length === 0) {
+      if (engineHotRelationship.length === 0) {
         engineRegistry.delete(engine);
+      }
+    }
+
+    if (sharedEngineUsage && sharedEngineUsage.includes(hotInstance.guid)) {
+      sharedEngineUsage.splice(sharedEngineUsage.indexOf(hotInstance.guid), 1);
+
+      if (sharedEngineUsage.length === 0) {
+        sharedEngineRegistry.delete(engine);
         engine.destroy();
       }
     }
```

Comment:
We can probably deduplicate those loops by extracting `deleteEngine` function 🤔

Reply:
I'll leave it for the next iteration.